### ISSUE TITLE
History run results are different from the initial run

### ISSUE THREAD

**Issue by hamid4965** [EDmodel/ED2#255]
I'm trying to run the ED in the history mode. To test, I compared initial run with history run. So first I ran the model in the "INITIAL" mode from 1/1/2009 to 31/12/2013. Then I ran ED in the "HISTORY" mode from 1/1/2012 to 31/12/2013. The ED2IN setting for history run is as follow: 

NL%RUNTYPE  = 'HISTORY'
NL%OUTFAST    = -1
NL%OUTSTATE   = 0
NL%FRQFAST   = 1800
NL%FRQSTATE  = 1
NL%IED_INIT_MODE   = 6
NL%ITIMEH   = 0000   ! UTC
NL%IDATEH   = 01     ! Day
NL%IMONTHH  = 01     ! Month
NL%IYEARH   = 2012   ! Year

Figure shows the comparison of NPP for 2012-2013 extracted from INITIAL simulations (Red color) and history simulations (Black color). As you can see they are very different. My question is should the history run exactly matches the initial simulations? Any idea what might be the reason? 

![rplot](https://user-images.githubusercontent.com/5567953/46112905-87fdef80-c1a9-11e8-938b-59948c76a22b.png)
Here is my ED2IN file
[edin_hist1.txt](https://github.com/EDmodel/ED2/files/2421770/edin_hist1.txt)

**Comment by mpaiao** [comment]
@hamid4965 I checked your ED2IN and I noticed that you are using PFT 18, and you are probably setting the PFT parameters in the xml file.  Because PFT 18 is not a default PFT, my hunch is that not all the parameters you need are initialized through XML.

You could test this by running the model without XML and use one of the standard PFTs, and repeat your test, just to see if the problem goes away. In case it does, then you may need to identify which parameters are missing, otherwise we can think of other tests.

**Comment by xiangtaoxu** [comment]
@hamid4965 I think the problem is that you should set IED_INIT_MODE to 4. Your value is 6, which means the model will read the .pss and .css file instead of the .h5 state file (see the code below). Therefore, your simulation is not a true history restart because it does not include the growth/mortality from 2009 to 2012.

https://github.com/EDmodel/ED2/blob/39a3ddd65035c3ba4aed05d9528a3bf2748cdb8b/ED/src/init/ed_init.F90#L406-L425

@mpaiao Maybe there should be a fatal_error for this scenario in ed_opspec.f90?

**Comment by mpaiao** [comment]
@xiangtaoxu This is a bit confusing, but RUNTYPE="HISTORY" and IED_INIT_MODE=4 do different things.

RUNTYPE="HISTORY": every variable in the -S- files is read, and the code should continue an existing simulation and produce the same results.  In this case, IED_INIT_MODE is ignored, and this option only makes sense if the goal is to continue a simulation that was interrupted with the exact same settings and exact same executable.

IED_INIT_MODE =4 (or 5), using RUNTYPE="INITIAL": ED will start the forest structure state from an existing h5 file, but it will not read everything (e.g. soil moisture, bstorage, etc. are not read)..  This is equivalent to a pss/css initialization, but reading from h5 instead of ascii.  When IED_INIT_MODE=4, the code must find a polygon that matches lon/lat with the polygon of interest, whereas IED_INIT_MODE=5 simply picks the closest polygon from all the files that match SFILIN.   Both options 4 and 5 should work even if the h5 files come from older versions of ED2 and different settings.

**Comment by xiangtaoxu** [comment]
@mpaiao I see the code in ed_driver.f90 now. This is a long-time mistake of mine... Thanks!

**Comment by hamid4965** [comment]
Thanks @mpaiao and @xiangtaoxu . Past few days I was running different tests. I should say that In my first post I used the older version of ED with no hydraulic module. Now I switch to the new one with hydraulic. 
1- @mpaiao  I'm working on a new pft (pft 18), but I defined it in ed_params.f90. I didn't use XML to define a new pft, instead, I used XML to change just one parameter (SLA). Anyway, I tried to run the model with old pfts (pft=8) and setting NL%IED_INIT_MODE   = 2 and no config.xml file. As you can see from the figure the problem still exists. 
![pft_8_hydro](https://user-images.githubusercontent.com/5567953/46261156-c5fe5a80-c4ac-11e8-9cda-e28dc5976d8d.png)
I'm not sure what I'm doing wrong. The ED2IN is attached: 

[ED2IN.txt](https://github.com/EDmodel/ED2/files/2431870/ED2IN.txt)

2- @xiangtaoxu,  I tried to run the model with NL%IED_INIT_MODE   = 6 and INITIAL mode but everything else similar to the above run and I got the following error which I don't really undrestand why: 

=== Time integration starts (model) ===
 - Simulating:   01/01/2009 00:00:00 UTC
|--------------------------------------------------|
|          !!! ): Water budget failed :( !!!       |
|--------------------------------------------------|
 TIME           : 2009 01 01     0.
 LAI            :   1.3442509E+00
 VEG_HEIGHT     :   5.7854939E-01
 CAN_DEPTH      :   5.0000000E+00
 RESIDUAL       :             NaN
 INITIAL_STORAGE:             NaN
 FINAL_STORAGE  :             NaN
 DELTA_STORAGE  :             NaN
 PRECIPGAIN     :   0.0000000E+00
 DENSITY_EFFECT :  -2.1347903E-09
 LOSS2ATM       :   2.1377521E-02
 LOSS2DRAINAGE  :   1.0416529E-03
 LOSS2RUNOFF    :   0.0000000E+00
|--------------------------------------------------|

According to ED2IN, IED_INIT_MODE option 6 "is exactly like option 2, except that the PFT types are preserved" So I don't know why this option should fail? 

All I'm trying to do is to spin up ED for say hundred years and use the history of this spin up run for next model runs with different settings. Any suggestion for ED setting in this regard would be really helpful. 

Thanks

**Comment by xiangtaoxu** [comment]
@hamid4965 I think the Water Budget problem under plant hydraulics run is because leaf/wood water content was not initiated properly likely due to erroneous parameterization. Can you upload a copy of the output xml file that records all the parameters in your simulation? It should be in your output folder with the same prefix as your other output files.

**Comment by mpaiao** [comment]
I agree with @xiangtaoxu that the water budget problem is likely an initial value bug.  Regarding the history runs, one suggestion is to go through the -S- and -Z- files at the time you start the simulation using RUNTYPE='HISTORY' (i.e. the time that matches IYEARH, IMONTHH, etc.).  The -S- file is the file generated in the first run (RUNTYPE = 'INITIAL'), and the Z file is generated after reading the -S- file.  If I remember correctly, they should be identical, so if you are able to spot which variables do not match, that could point to the places to look for.  There are lots of variables, so you may need to code some script that searches for differences.

**Comment by crollinson** [comment]
This looks like it could be similar to issues I was having with initialization earlier (see #252 and #251).  If you're running with hand-me-down Dietze lab ED2INs & .css/.pss files there are some changes that need to be made.

**Comment by xiangtaoxu** [comment]
@hamid4965 Great to know it is solved. A side note about rho. The model does read rho values from xml but will not re-calculate all the variables that are dependent on rho. You have to manually change the hydraulic parameters in the xml at the same time.
